File: lab6/task4/go_solution/t4.go
package main

import (
	"fmt"
	"os"
)

var N, M int

// Функция для красивого вывода маски в виде доски в терминал
func printMatrix(mask int) {
	for i := 0; i < N; i++ {
		fmt.Print("  ")
		for j := 0; j < M; j++ {
			bit := (mask >> (i*M + j)) & 1
			fmt.Printf("%d ", bit)
		}
		fmt.Println()
	}
}

// Генерация целевых состояний "Рамки"
func getTargetMasks() (int, int) {
	maskA := 0 // По краю 1, внутри 0
	maskB := 0 // По краю 0, внутри 1

	for i := 0; i < N; i++ {
		for j := 0; j < M; j++ {
			bitPos := i*M + j
			isBorder := (i == 0 || i == N-1 || j == 0 || j == M-1)

			if isBorder {
				maskA |= (1 << bitPos)
			} else {
				maskB |= (1 << bitPos)
			}
		}
	}
	return maskA, maskB
}

func solve(startMask int) int {
	targetA, targetB := getTargetMasks()

	// Если стартовое состояние уже является рамкой
	if startMask == targetA || startMask == targetB {
		return 0
	}

	// Очередь для BFS (используем slice)
	queue := []int{startMask}

	// Карта расстояний (хранит {состояние : количество_ходов})
	dist := make(map[int]int)
	dist[startMask] = 0

	for len(queue) > 0 {
		// Достаем элемент из начала очереди
		curr := queue[0]
		queue = queue[1:]

		currDist := dist[curr]

		// Перебираем все пары соседних клеток
		for i := 0; i < N; i++ {
			for j := 0; j < M; j++ {
				p1 := i*M + j

				// Сосед справа
				if j+1 < M {
					p2 := i*M + (j + 1)
					nextMask := curr ^ (1 << p1) ^ (1 << p2)

					if nextMask == targetA || nextMask == targetB {
						return currDist + 1
					}
					if _, visited := dist[nextMask]; !visited {
						dist[nextMask] = currDist + 1
						queue = append(queue, nextMask)
					}
				}

				// Сосед снизу
				if i+1 < N {
					p2 := (i+1)*M + j
					nextMask := curr ^ (1 << p1) ^ (1 << p2)

					if nextMask == targetA || nextMask == targetB {
						return currDist + 1
					}
					if _, visited := dist[nextMask]; !visited {
						dist[nextMask] = currDist + 1
						queue = append(queue, nextMask)
					}
				}
			}
		}
	}
	return -1
}

func main() {
	fmt.Println("=========================================================")
	fmt.Println("   ПРОГРАММА ПОИСКА МИНИМАЛЬНОГО ПУТИ К РАМКЕ МОНЕТ (GO)")
	fmt.Println("=========================================================")
	fmt.Println()

	fmt.Println("[Шаг 1] Введите размеры щита (N - строки, M - столбцы):")
	fmt.Print("Формат ввода: N M (например, 3 3): ")

	_, err := fmt.Scan(&N, &M)
	if err != nil || N <= 0 || M <= 0 {
		fmt.Println("\n[Ошибка] Некорректный размер доски!")
		os.Exit(1)
	}

	// Проверка ограничений для битовой маски (32-битный int)
	if N*M > 30 {
		fmt.Println("\n[Предупреждение] Размер доски слишком велик для BFS перебора масок (> 30 клеток).")
		os.Exit(1)
	}

	// Показываем пользователю целевые конфигурации
	tA, tB := getTargetMasks()
	fmt.Printf("\n[Инфо] Целевые состояния («Рамки») для размера %dx%d:\n", N, M)
	fmt.Println("Вариант А (1 по краям):")
	printMatrix(tA)
	fmt.Println("Вариант Б (0 по краям):")
	printMatrix(tB)
	fmt.Println("---------------------------------------------------------")
	fmt.Println()

	fmt.Println("[Шаг 2] Введите матрицу монет (строка за строкой).")
	fmt.Println("Используйте '0' для орла и '1' для решки:")

	startMask := 0
	for i := 0; i < N; i++ {
		fmt.Printf("Строка %d (введите %d чисел через пробел): ", i+1, M)
		for j := 0; j < M; j++ {
			var val int
			fmt.Scan(&val)
			if val != 0 && val != 1 {
				fmt.Println("\n[Ошибка] Разрешены только значения 0 или 1!")
				os.Exit(1)
			}
			if val == 1 {
				startMask |= (1 << (i*M + j))
			}
		}
	}

	fmt.Println("\n---------------------------------------------------------")
	fmt.Println("[Статус] Исходная доска принята. Запускаю расчет алгоритма BFS...")
	fmt.Println("---------------------------------------------------------")
	fmt.Println()

	result := solve(startMask)

	fmt.Println("==================== РЕЗУЛЬТАТ ==========================")
	if result == -1 {
		fmt.Println(" Исходное состояние: ")
		printMatrix(startMask)
		fmt.Println("\n ПРИГОВОР: Из данного расположения получить рамку НЕВОЗМОЖНО.")
	} else {
		fmt.Println(" УСПЕХ! Целевой рисунок достижим.")
		fmt.Printf(" Минимальное количество ходов (переворотов пар): %d\n", result)
	}
	fmt.Println("=========================================================")
}

File: lab6/task4/cpp_solution/t4.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>

using namespace std;

int N, M;

// Функция для красивого вывода маски в виде доски в терминал
void print_matrix(int mask) {
    for (int i = 0; i < N; ++i) {
        cout << "  ";
        for (int j = 0; j < M; ++j) {
            int bit = (mask >> (i * M + j)) & 1;
            cout << bit << " ";
        }
        cout << "\n";
    }
}

// Генерация целевых состояний
pair<int, int> get_target_masks() {
    int maskA = 0; // По краю 1, внутри 0
    int maskB = 0; // По краю 0, внутри 1
    
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < M; ++j) {
            int bit_pos = i * M + j;
            bool is_border = (i == 0 || i == N - 1 || j == 0 || j == M - 1);
            
            if (is_border) {
                maskA |= (1 << bit_pos);
            } else {
                maskB |= (1 << bit_pos);
            }
        }
    }
    return {maskA, maskB};
}

int solve(int start_mask) {
    auto [targetA, targetB] = get_target_masks();
    
    if (start_mask == targetA || start_mask == targetB) {
        return 0;
    }
    
    queue<int> q;
    unordered_map<int, int> dist;
    
    q.push(start_mask);
    dist[start_mask] = 0;
    
    while (!q.empty()) {
        int curr = q.front();
        q.pop();
        int curr_dist = dist[curr];
        
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < M; ++j) {
                int p1 = i * M + j;
                
                // Сосед справа
                if (j + 1 < M) {
                    int p2 = i * M + (j + 1);
                    int next_mask = curr ^ (1 << p1) ^ (1 << p2);
                    if (next_mask == targetA || next_mask == targetB) return curr_dist + 1;
                    if (dist.find(next_mask) == dist.end()) {
                        dist[next_mask] = curr_dist + 1;
                        q.push(next_mask);
                    }
                }
                // Сосед снизу
                if (i + 1 < N) {
                    int p2 = (i + 1) * M + j;
                    int next_mask = curr ^ (1 << p1) ^ (1 << p2);
                    if (next_mask == targetA || next_mask == targetB) return curr_dist + 1;
                    if (dist.find(next_mask) == dist.end()) {
                        dist[next_mask] = curr_dist + 1;
                        q.push(next_mask);
                    }
                }
            }
        }
    }
    return -1;
}

int main() {
    // Настройка локализации для корректного вывода кириллицы
    setlocale(LC_ALL, "Russian");

    cout << "=========================================================\n";
    cout << "   ПРОГРАММА ПОИСКА МИНИМАЛЬНОГО ПУТИ К РАМКЕ МОНЕТ\n";
    cout << "=========================================================\n\n";

    cout << "[Шаг 1] Введите размеры щита (N - строки, M - столбцы):\n";
    cout << "Формат ввода: N M (например, 3 3): ";
    if (!(cin >> N >> M) || N <= 0 || M <= 0) {
        cout << "\n[Ошибка] Некорректный размер доски!\n";
        return 1;
    }

    // Проверка ограничений для битовой маски (максимум 30-32 бита для стандартного int)
    if (N * M > 30) {
        cout << "\n[Предупреждение] Размер доски слишком велик для BFS перебора масок (> 30 клеток).\n";
        return 1;
    }

    // Показываем пользователю, какие рамки будут считаться целью
    auto [tA, tB] = get_target_masks();
    cout << "\n[Инфо] Целевые состояния («Рамки») для размера " << N << "x" << M << ":\n";
    cout << "Вариант А (1 по краям):\n";
    print_matrix(tA);
    cout << "Вариант Б (0 по краям):\n";
    print_matrix(tB);
    cout << "---------------------------------------------------------\n\n";

    cout << "[Шаг 2] Введите матрицу монет (строка за строкой).\n";
    cout << "Используйте '0' для орла и '1' для решки:\n";
    
    int start_mask = 0;
    for (int i = 0; i < N; ++i) {
        cout << "Строка " << i + 1 << " (введите " << M << " чисел через пробел): ";
        for (int j = 0; j < M; ++j) {
            int val;
            cin >> val;
            if (val != 0 && val != 1) {
                cout << "\n[Ошибка] Разрешены только значения 0 или 1!\n";
                return 1;
            }
            if (val == 1) {
                start_mask |= (1 << (i * M + j));
            }
        }
    }

    cout << "\n---------------------------------------------------------\n";
    cout << "[Статус] Исходная доска принята. Запускаю расчет алгоритма BFS...\n";
    cout << "---------------------------------------------------------\n\n";

    int result = solve(start_mask);

    cout << "==================== РЕЗУЛЬТАТ ==========================\n";
    if (result == -1) {
        cout << " Исходное состояние: \n";
        print_matrix(start_mask);
        cout << "\n ПРИГОВОР: Из данного расположения получить рамку НЕВОЗМОЖНО.\n";
    } else {
        cout << " УСПЕХ! Целевой рисунок достижим.\n";
        cout << " Минимальное количество ходов (переворотов пар): " << result << "\n";
    }
    cout << "=========================================================\n";

    return 0;
}